### PR TITLE
Add hierarchical accumulation

### PR DESCRIPTION
**Pull request by hkctkuy**

### CODE REVIEW

File: libcasr/src/stacktrace.rs [hunk 1]
```diff
@@ -189,30 +189,21 @@ pub fn dedup_stacktraces(stacktraces: &[Stacktrace]) -> Vec<bool> {
...
 
-/// Perform the clustering of stack traces
+/// Perform the clustering by condensed dissimilarity matrix
 ///
 /// # Arguments
 ///
-/// * `stacktraces` - slice of `Stacktrace` structures
+/// * `len` - matrix length
+///
+/// * `matrix` - condensed dissimilarity matrix
```
Comment (SweetVishnya) [inline libcasr/src/stacktrace.rs:198]: different arguments order in docs

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: libcasr/src/stacktrace.rs (hunk 1 above)
Comment (hkctkuy) [reply libcasr/src/stacktrace.rs:196]: Fixed

**Review by SweetVishnya — APPROVED** [review]

File: libcasr/src/stacktrace.rs [hunk 2]
```diff
@@ -189,30 +189,21 @@ pub fn dedup_stacktraces(stacktraces: &[Stacktrace]) -> Vec<bool> {
...
 
-/// Perform the clustering of stack traces
+/// Perform the clustering by condensed dissimilarity matrix
 ///
 /// # Arguments
 ///
-/// * `stacktraces` - slice of `Stacktrace` structures
+/// * `matrix` - condensed dissimilarity matrix
+///
+/// * `len` - matrix length
```
Comment (Avgor46) [inline libcasr/src/stacktrace.rs:198]: number of observations that are being clustered

File: casr/src/bin/casr-cluster.rs [hunk 3]
```diff
@@ -401,94 +401,131 @@ fn update_clusters(
...
+
+    // Sync real cluster numbers with resulting numbers
+    let mut numbers: HashMap<usize, usize> = HashMap::new();
+    for i in 0..clusters.len() {
+        numbers.insert(res[i], *keys[i]);
+    }
+    // New clusters
+    let mut news: HashMap<usize, Cluster> = HashMap::new();
+    let mut new_num = max;
+    for &num in res.iter().skip(clusters.len()).take(deviants.len()) {
```
[the remark this answers is not in the document]
Comment (hkctkuy) [reply casr/src/bin/casr-cluster.rs]: Removed

File: libcasr/src/cluster.rs [hunk 4]
```diff
@@ -186,32 +186,26 @@ impl Cluster {
...
-                if number == i {
-                    crashline = line.to_string();
-                    break;
-                }
+    /// Get complete distance between cluster and report
+    /// NOTE: Result also can be interpreted as diameter of cluster merge result
+    pub fn dist_rep(cluster: &Cluster, report: &ReportInfo) -> f64 {
+        let mut max = 0f64;
+        let (_, (trace, _)) = report;
+        for stacktrace in cluster.stacktraces() {
```
[the remark this answers is not in the document]
Comment (hkctkuy) [reply libcasr/src/cluster.rs]: Fixed

**Review by Avgor46 — APPROVED** [review]

**Review by SweetVishnya — APPROVED** [review]

**Pull request merged**